portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -36,18 +42,56 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-    ArrayList<String> messages = new ArrayList<>();
+    HashMap<Integer, String> cursors = new HashMap<Integer, String>();
+    int pageCount = 0;
+    
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    ArrayList<String> messages = new ArrayList<>();
+
+    String nextPage = request.getParameter("next");
+    String cursor = request.getParameter("cursor"); 
+    pageCount += 1;
+    if(cursor != null && nextPage.equals("false")){
+        pageCount -= 2;
+        cursor = cursors.get(pageCount);
+    }
+
+    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(5);
+    if(cursor != null){
+
+        fetchOptions.startCursor(Cursor.fromWebSafeString(cursor));
+    } else {
+        cursors.put(pageCount, "");
+    }
+
+
+
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
     Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
-    PreparedQuery results = datastore.prepare(query);
-    for (Entity entity : results.asIterable()) {
-        Object content = entity.getProperty("content");
-        Object timestamp = entity.getProperty("timestamp");
-        Comment c = new Comment("None", content, timestamp);
-        messages.add(commentToJson(c));
+    if(nextPage == "false"){
+        query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
+    }
+    PreparedQuery resultsPQ = datastore.prepare(query);
```

Comment:
Typical style guidelines are to name this resultsPreparedQuery. Longer variable names are OK :).